pkg/chunk/purger/purger_test.go

```diff
@@ -377,6 +387,87 @@ func TestDataPurger_Restarts(t *testing.T) {
+
+	require.Equal(t, float64(1), testutil.ToFloat64(newPurger.metrics.deleteRequestsProcessedTotal))
+	require.PanicsWithError(t, "collected 0 metrics instead of exactly 1", func() {
+		testutil.ToFloat64(newPurger.metrics.deleteRequestsProcessingFailures)
+	})
+}
+
+func TestPurger_Metrics(t *testing.T) {
+	deleteStore, chunkStore, _, purger, registry := setupStoresAndPurger(t)
+	defer func() {
+		purger.StopAsync()
+		chunkStore.Stop()
+	}()
+
+	// start loop to load requests
+	require.NoError(t, services.StartAndAwaitRunning(context.Background(), purger))
+
+	// no delete requests for processing so age and pending request is 0 while we have successfully attempted loading request once.
+	require.NoError(t, testutil.GatherAndCompare(registry, strings.NewReader(
+		`
+			# HELP cortex_purger_load_pending_requests_attempts_total Number of attempts that were made to load pending requests with status
+			# TYPE cortex_purger_load_pending_requests_attempts_total counter
+			cortex_purger_load_pending_requests_attempts_total{status="success"} 1
+			# HELP cortex_purger_oldest_pending_delete_request_age_seconds Age of oldest pending delete request in seconds
+			# TYPE cortex_purger_oldest_pending_delete_request_age_seconds gauge
+			cortex_purger_oldest_pending_delete_request_age_seconds 0
+			# HELP cortex_purger_pending_delete_requests_count Count of requests which are in process or are ready to be processed
+			# TYPE cortex_purger_pending_delete_requests_count gauge
+			cortex_purger_pending_delete_requests_count 0
+		`),
+		"cortex_purger_load_pending_requests_attempts_total",
+		"cortex_purger_oldest_pending_delete_request_age_seconds",
+		"cortex_purger_pending_delete_requests_count",
+	))
+
+	// add delete request whose createdAt is now
+	err := deleteStore.AddDeleteRequest(context.Background(), userID, model.Time(0).Add(24*time.Hour),
+		model.Time(0).Add(2*24*time.Hour), []string{"foo"})
+	require.NoError(t, err)
+
+	// add delete request whose createdAt is 2 days back
+	err = deleteStore.addDeleteRequest(context.Background(), userID, model.Now().Add(-2*24*time.Hour), model.Time(0).Add(24*time.Hour),
+		model.Time(0).Add(2*24*time.Hour), []string{"foo"})
+	require.NoError(t, err)
+
+	// add delete request whose createdAt is 3 days back
+	err = deleteStore.addDeleteRequest(context.Background(), userID, model.Now().Add(-3*24*time.Hour), model.Time(0).Add(24*time.Hour),
+		model.Time(0).Add(8*24*time.Hour), []string{"foo"})
+	require.NoError(t, err)
+
+	// load new delete requests for processing
+	require.NoError(t, purger.pullDeleteRequestsToPlanDeletes())
+
+	// there must be 2 pending delete requests, oldest being 3 days old
+	require.InDelta(t, float64(3*86400), testutil.ToFloat64(purger.metrics.oldestPendingDeleteRequestAgeSeconds), 1)
+	require.Equal(t, float64(2), testutil.ToFloat64(purger.metrics.pendingDeleteRequestsCount))
+
+	// wait until purger_delete_requests_processed_total starts to show up.
+	for {
```

Comment:
Instead of this `for` loops, could you use `test.Poll()`? The advantage of `test.Pool()` is that it has a timeout too.

_Same comment applies to the next `for`_

Reply:
nice, I didn't know it existed. Changed it, thanks!